cellmlmanip/model.py

```diff
@@ -619,6 +619,205 @@ def variables(self):
         """ Returns an iterator over this model's variable symbols. """
         return self._name_to_symbol.values()
 
+    def add_input(self, name, units, is_output=False):
+        """ Adds an additional input for the variable name to the model with new units
+        and apply any necessary conversions/add additional equations.
+
+        For example:
+        Original model
+            x = 1 [pA]
+              in [pA]
+              oxmeta: current
+            y = 1 / x
+              in [1/pA]
+
+        add_input('x', 'nA') creates model
+            input_current = 1e-3 [nA]
+              in [nA]
+              oxmeta: current
+            x = input_current * 1e3 [pA/nA]
+              in [pA]
+            y = 1 / x
+              in [1/pA]
+
+        :param name: name of variable whose units are to be changed
+        :param units: A `pint` units representation
+        :param is_output: the variable is an output rather than an input
+        """
+        # check units is of type Unit
+        assert isinstance(units, self.units.ureg.Unit)
+
+        # check variable is in model
+        original_variable = self.get_symbol_by_name(name)
+        if not original_variable:
+            raise KeyError('No variable with name "%s" found.' % name)
+
+        original_units = original_variable.units
+
+        # no conversion necessary
+        if original_units == units:
+            return
+
+        state_symbols = self.get_state_symbols()
+        free_symbol = self.get_free_variable_symbol()
+        # conversion_factor for old units to new units
+        cf = self.units.get_conversion_factor(from_unit=original_units, to_unit=units)
+
+        # create new variable and relevant equations
+        new_variable = self._convert_variable_instance(original_variable, cf, units)
+
+        # if is output do not need to do additional changes for state/free symbols
+        if is_output:
+            return
+
+        new_derivatives = []
+        # if state variable
+        if original_variable in state_symbols:
+            new_derivatives.append(self._convert_state_variable_deriv(original_variable, new_variable, cf))
+
+        # if free variable
+        if original_variable == free_symbol:
+            # for each derivative wrt to free variable add necessary variables/equations
+            current_equations = self.equations.copy()
+            for equation in current_equations:
+                if equation.args[0].is_Derivative:
+                    if equation.args[0].args[1].args[0] == original_variable:
+                        new_derivatives.append(self._convert_free_variable_deriv(equation, new_variable, cf))
+
+        # replace any instances of derivative of rhs of other eqns with new derivative variable
+        for new_derivative in new_derivatives:
+            self._replace_derivatives(new_derivative)
```

Comment:
Incidentally, will `self.get_state_symbols` give the right list after we've done this? Should make sure to test that.

Reply:
So if we have state_symbol sv1 in mV and we do add_input(sv1, V) we will get a new variable named 'sv1_converted' which is now the state symbol. Just checking I will still have same number of state symbols, but original symbol_sv1 will no longer be a state var but the new symbol_sv1_converted will be

Reply:
This is what current code does - is it right ?

Reply:
Yes, that's correct - if the flow direction is input. If output, then the original state var stays a state var.